### PR TITLE
Make non-existent repos require login to clone

### PR DESCRIPTION
**Pull request by gary-kim**
Previously, it was possible to check if a private
repository exists by attempting to clone the repo.

This change will make it so that non-existent repos
will also require a login and will only tell the user
whether the truly doesn't if they would know about a
private repository with that name.

This will only really be effective while Gitea is in prod mode. Otherwise, errors that reveal exactly why the user could not clone may be sent to the user.

EDIT: I'll see if I can add some tests.

### CODE REVIEW

**Comment by guillep2k** [comment]
Can this be implemented in `routers.RegisterRoutes()` with a filter function instead?

**Comment by gary-kim** [comment]
> Can this be implemented in `routers.RegisterRoutes()` with a filter function instead?

I think I could turn it into a middleware function. Give me a bit.

**Comment by lunny** [comment]
@lafriks why this is a security related problem? I think the original implementation is security but not friendly for git.

**Comment by lafriks** [comment]
You can check that such repository exists even if you don't know that

**Comment by guillep2k** [comment]
Perhaps I'm missing something, but I see a lot of auth logic in the PR. Shouldn't that logic exist already elsewhere?

**Comment by gary-kim** [comment]
@guillep2k Yes, it already existed. I moved the auth logic into a function so that it can be reused so those changes are showing up in the PR.

File: routers/repo/http.go [hunk 1]
```diff
@@ -122,15 +127,81 @@ func HTTP(ctx *context.Context) {
...
+	if setting.Service.EnableReverseProxyAuth && len(authUsername) > 0 {
+		authUser, err = models.GetUserByName(authUsername)
+		if err != nil {
+			ctx.HandleText(401, "reverse proxy login error, got error while running GetUserByName")
+			return
+		}
+	} else {
+		authHead := ctx.Req.Header.Get("Authorization")
+		if len(authHead) == 0 {
+			ctx.Resp.Header().Set("WWW-Authenticate", "Basic realm=\".\"")
```
Comment (guillep2k) [inline routers/repo/http.go:201]: Very strange. This line shows up as brand new from your code, except it isn't! It's in the original `http.go` file. I should see a "diff minus" section with these lines if for example you moved them, or you changed the indentation. How can that be? I can see your branch is correctly based on the current master, so I'm puzzled.

image

That's what caught my attention about you "implementing" a lot of auth stuff.

**Comment by Zettat123** [comment]
Looks like this issue has been resolved after #8419 was merged?

**Pull request closed without merging**